extensions/vertx-web/deployment/src/test/java/io/quarkus/vertx/web/reactive/RequestContextPropagationTest.java

```diff
@@ -0,0 +1,75 @@
+package io.quarkus.vertx.web.reactive;
+
+import static io.restassured.RestAssured.get;
+import static org.hamcrest.Matchers.is;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import javax.annotation.PreDestroy;
+import javax.enterprise.context.RequestScoped;
+import javax.inject.Inject;
+import javax.inject.Singleton;
+
+import org.jboss.shrinkwrap.api.ShrinkWrap;
+import org.jboss.shrinkwrap.api.spec.JavaArchive;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+import io.quarkus.test.QuarkusUnitTest;
+import io.quarkus.vertx.web.Route;
+import io.smallrye.mutiny.Uni;
+import io.vertx.ext.web.RoutingContext;
+
+public class RequestContextPropagationTest {
+
+    @RegisterExtension
+    static final QuarkusUnitTest config = new QuarkusUnitTest()
+            .setArchiveProducer(() -> ShrinkWrap.create(JavaArchive.class)
+                    .addClasses(MyRoutes.class, Ping.class));
+
+    @Test
+    public void test() {
+        get("/ping?val=bar").then().statusCode(200).body(is("foo_bar"));
+        assertTrue(Ping.DESTROYED.get());
+    }
+
+    @Singleton
+    static class MyRoutes {
+
+        @Inject
+        Ping ping;
+
+        @Route(path = "ping")
+        void ping(RoutingContext ctx) {
+            // Init the Ping bean
+            ping.init("foo_");
+            Uni.createFrom().item(ctx.request().getParam("val")).subscribe().with(
```

Comment:
Did you actually run this test without your fix? Did it fail?
I'm asking because this looks like a synchronous call: there's nothing that would make it switch thread or delay execution.
If it does already pass without your fix (as I think it should), try switching your code to:

```java
Uni.createFrom().item(ctx.request().getParam("val")).onItem().delayIt().by(Duration.ofMillis(200)).subscribe().with(...)
```

Reply:
Yes, I did and it failed. Most probably because my machine is fast and the `ping()` method completes before the `Uni` callback. `.delayIt().by(Duration.ofMillis(200))` sounds reasonable though. Thanks...

Reply:
The speed of your machine should not be able to explain it failing. Either `Uni.createFrom().item(ctx.request().getParam("val")).subscribe().with()` is synchronous, or it's not. If it's synchronous your callback will be called before the method returns and before you get into your interceptor, which therefore should not fail the test because we haven't destroyed the context yet.

@cescoffier where am I wrong here?

Reply:
Hm, why do you think it's synchronous? I'm definitely no expert here but I'd expect it to be async... @cescoffier will tell us ;-).

Reply:
Well, `item()` is synchronous, no?

Reply:
Yes, but `Uni.subscribe()` is not... but I may be completely wrong ;-).

Reply:
The `item` method will emit the item immediately after subscription in a synchronous fashion (using the thread used for the subscription).

 You can simulate some delay with `delayIt` as pointed by @FroMage.

Reply:
Just to clarify:

```
Uni.createFrom().item("hello").subscribe().with(t -> Systme.out.println(t));
System.out.println("Done");
```

prints:

```
hello
Done
```

```
Uni.createFrom().item("hello")
    .onItem().delayIt().by(Duration.ofSecond(1))
    .subscribe().with(t -> Systme.out.println(t));
System.out.println("Done");
```

prints:

```
Done
hello # printed roughly 1 second later
```

Reply:
Ok, I should double check my test then... Anyway, from my "reactive noob" POV it feels weird that `onItem().delayIt().by(Duration.ofSecond(1))` causes async execution. I read the javadoc of `subscribe()` as a non-blocking subscription...

Reply:
Well, subscribe depends on whether the left-hand-side is already completed or not. In the case of a single item it is, so synchronous. In the case of delay, it's not, so it will be completed later, so async.
Now the much bigger questions is how could this test possibly fail before you put the fix, since it's all synchronous? That's what's worrying me.

Reply:
>  That's what's worrying me.

Maybe it's just a PEBKAC on my side ;-). I will tell you in few mins...

Reply:
`subscribe` is non-blocking. It does not block the caller thread. It receives the item once it is emitted. 

So the interesting aspect is _when_ your item is emitted. In your code, the item is emitted during the subscription (immediately). That's what `Uni.createFrom().item()` is doing. Then, every processing stage is synchronous in your code - which means each stage is called immediately (well, you don't do any processing so it's obviously synchronous). That explains why the subscriber is called directly (on the subscriber thread).

Reply:
Good news everyone. It was PEBKAC... I wasn't able to reproduce the failure without `onItem().delayIt()`. Thanks for clarification guys!

Reply:
OK, pfew, at least we are more confident in the test and fix now :)